kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/SchemaUtils.java

```diff
@@ -32,6 +33,28 @@ public class SchemaUtils {
 
   private SchemaUtils() {}
 
+  /** Checks if the {@code supersetSchema} is a superset of the {@code subsetSchema}. */
+  public static boolean isSuperset(StructType supersetSchema, StructType subsetSchema) {
+    final Map<String, StructField> supersetFieldsMap =
+        flattenNestedFields(supersetSchema).stream()
+            .collect(Collectors.toMap(x -> x._1, x -> x._2));
+    final List<Tuple2<String, StructField>> subsetNameFields = flattenNestedFields(subsetSchema);
+
+    for (Tuple2<String, StructField> subsetNameField : subsetNameFields) {
+      if (!supersetFieldsMap.containsKey(subsetNameField._1)) {
+        return false;
+      }
+
+      final StructField supersetField = supersetFieldsMap.get(subsetNameField._1);
+
+      if (!supersetField.getDataType().equivalent(subsetNameField._2.getDataType())) {
+        return false;
+      }
```

Comment:
Please correct me if I'm wrong, but I think (based on this + I don't think I see a test for it) we will fail if you try to prune any nested columns right?

For example if we have table schema
```
| - a: INTEGER
| - b: STRUCT
|     | - 1: INTEGER
|     | - 2: STRING
```

We should allow a read schema of 
```
| - a: INTEGER
| - b: STRUCT
|     | - 1: INTEGER
```

I think we should allow this? For the purposes of this check I think we only want to check for matching leaf columns and not the non-leaf nodes.

Reply:
I was surprised that `flattenNestedFieldNames` includes leaf and non-leaf columns I would've expected it to be only the leaf columns. Not sure if it was implemented that way originally for the other place it is used?

Reply:
But maybe also we could support this by making this check here in line 50 a recursive `isSuperSet` check? idk there's multiple ways we could support it

Reply:
Or I think in that case maybe we don't flatten everything and we just do recursive `isSuperSet` checks on all fields?

Reply:
Great callout! Let me re-implement and re-test this.

Reply:
Pushed some changes that resolves this